src/stream/src/executor/over_window/general.rs

```diff
@@ -0,0 +1,925 @@
+        })
+    }
+}
+
+pub struct OverWindowExecutorArgs<S: StateStore> {
+    pub input: BoxedExecutor,
+
+    pub actor_ctx: ActorContextRef,
+    pub pk_indices: PkIndices,
+    pub executor_id: u64,
+
+    pub calls: Vec<WindowFuncCall>,
+    pub partition_key_indices: Vec<usize>,
+    pub order_key_indices: Vec<usize>,
+    pub order_key_order_types: Vec<OrderType>,
+
+    pub state_table: StateTable<S>,
+    pub watermark_epoch: AtomicU64Ref,
+
+    pub chunk_size: usize,
+}
+
+impl<S: StateStore> OverWindowExecutor<S> {
+    pub fn new(args: OverWindowExecutorArgs<S>) -> Self {
+        let input_info = args.input.info();
+
+        let schema = {
+            let mut schema = input_info.schema.clone();
+            args.calls.iter().for_each(|call| {
+                schema.fields.push(Field::unnamed(call.return_type.clone()));
+            });
+            schema
+        };
+
+        Self {
+            input: args.input,
+            inner: ExecutorInner {
+                actor_ctx: args.actor_ctx,
+                info: ExecutorInfo {
+                    schema,
+                    pk_indices: args.pk_indices,
+                    identity: format!("OverWindowExecutor {:X}", args.executor_id),
+                },
+                calls: args.calls,
+                partition_key_indices: args.partition_key_indices,
+                order_key_indices: args.order_key_indices,
+                order_key_order_types: args.order_key_order_types,
+                input_pk_indices: input_info.pk_indices,
+                input_schema_len: input_info.schema.len(),
+                state_table: args.state_table,
+                watermark_epoch: args.watermark_epoch,
+                chunk_size: args.chunk_size,
+            },
+        }
+    }
+
+    async fn ensure_partition_in_cache(
+        this: &mut ExecutorInner<S>,
+        cache: &mut PartitionCache,
+        partition_key: &OwnedRow,
+    ) -> StreamExecutorResult<()> {
+        if cache.contains(partition_key) {
+            return Ok(());
+        }
+
+        let mut cache_for_partition = BTreeMap::new();
+        let table_iter = this
+            .state_table
+            .iter_with_pk_prefix(partition_key, PrefetchOptions::new_for_exhaust_iter())
+            .await?;
+
+        #[for_await]
+        for row in table_iter {
+            let row: OwnedRow = row?;
+            cache_for_partition.insert(this.row_to_state_key(&row)?, row);
+        }
+
+        cache.put(partition_key.clone(), Partition::new(cache_for_partition));
+        Ok(())
+    }
+
+    /// Merge changes by input pk in the given chunk, return a change iterator which guarantees that
+    /// each pk only appears once. This method also validates the consistency of the input
+    /// chunk.
+    ///
+    /// TODO(rc): We may want to optimize this by handling changes on the same pk during generating
+    /// partition [`Change`]s.
+    fn merge_changes_in_chunk<'a>(
```

Comment:
the logic is a little same with `MaterializeBuffer`

Reply:
Maybe we can leave this for now, I'm thinking that this function can possibly be optimized later.